library/alloc/src/vec.rs

```diff
@@ -2391,11 +2384,24 @@ where
 }
 
 impl<T> SpecExtend<T, IntoIter<T>> for Vec<T> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
+    fn spec_extend(&mut self, iterator: IntoIter<T>) {
+        // Avoid reallocation if we can use iterator's storage instead. This requires 1 memcpy and 0-1 memmove
```

Comment:
An ASCII diagram of how the contents of `self` end up prepended to `iterator`'s storage would make this a bit clearer.

Reply:
I still think it would be good to show the different cases and number of `memmove` and `memcpy`, it is very nice to say how many `memcpy` are there.

Reply:
Add a diagram. It's a bit meager since I tried to cram each step into 1 line.
If you want something more detailed then an example of what you have in mind would help.

Reply:
Pushed another version of the diagram.

Reply:
Nice diagram.